File: ui/supply_map.js
// Animated supply map for mfa-engine (Streamlit custom component, v2 API).
//
// Everything is prepared in Python (ui/supply_map.py): projected country outlines, circle sizes,
// arc geometry and the cascade rounds. This file only draws and animates, with plain SVG and the
// Web Animations API, no libraries. Motion is decoration, never information: with
// prefers-reduced-motion the map is drawn in its final state and the particles stay off.

const NS = "http://www.w3.org/2000/svg";

function svgEl(tag, attrs, parent) {
  const e = document.createElementNS(NS, tag);
  for (const k in attrs || {}) e.setAttribute(k, attrs[k]);
  if (parent) parent.appendChild(e);
  return e;
}

function htmlEl(tag, cls, parent, text) {
  const e = document.createElement(tag);
  if (cls) e.className = cls;
  if (text != null) e.textContent = text;
  if (parent) parent.appendChild(e);
  return e;
}

const ICONS = {
  play: "M8 5.5v13l11-6.5z",
  pause: "M7 5h4v14H7zM13 5h4v14h-4z",
  reset: "M12 5V2L7 6l5 4V7a5 5 0 1 1-5 5H5a7 7 0 1 0 7-7z",
};

function button(parent, label, icon, cls) {
  const b = htmlEl("button", "sm-btn" + (cls ? " " + cls : ""), parent);
  b.type = "button";
  const s = svgEl("svg", { viewBox: "0 0 24 24", width: "16", height: "16", "aria-hidden": "true" }, b);
  const p = svgEl("path", { d: ICONS[icon], fill: "currentColor" }, s);
  htmlEl("span", null, b, label);
  b.setIcon = (name) => p.setAttribute("d", ICONS[name]);
  b.setLabel = (text) => { b.querySelector("span").textContent = text; };
  return b;
}

// Point on a quadratic Bezier arc at parameter t.
function bez(f, t) {
  const u = 1 - t;
  return [u * u * f.x1 + 2 * u * t * f.cx + t * t * f.x2, u * u * f.y1 + 2 * u * t * f.cy + t * t * f.y2];
}

export default function mount(component) {
  const { data, parentElement } = component;
  if (!data || !data.map) return;
  if (parentElement.__smCleanup) {
    try { parentElement.__smCleanup(); } catch (err) { /* previous render already gone */ }
  }

  const reduce = window.matchMedia("(prefers-reduced-motion: reduce)").matches;
  const host = parentElement.querySelector(".sm-mount") || parentElement;
  const root = htmlEl("div", "sm-root", host);
  // Tell the page the map has drawn: the space held for it while loading (see ui/theme.py) is
  // released, and the element takes the map's own height, whatever the legend wraps to.
  const hostEl = parentElement.host || parentElement;
  const holder = hostEl.closest ? hostEl.closest('[data-testid="stElementContainer"]') : null;
  if (holder) holder.dataset.smReady = "1";
  root.style.setProperty("--acc", data.style.accent);
  root.style.setProperty("--acc-rgb", data.style.accentRgb);
  root.style.setProperty("--crit", data.style.critical);

  const timers = [];
  const later = (fn, ms) => { const id = setTimeout(fn, ms); timers.push(id); return id; };
  let raf = 0, running = false, flowsOn = !reduce, visible = true, last = 0;
  let pinned = null, played = false;

  // ------------------------------------------------------------------ toolbar
  const bar = htmlEl("div", "sm-bar", root);
  const playBtn = button(bar, "Play the disruption", "play", "primary");
  const flowBtn = button(bar, reduce ? "Show moving flows" : "Pause flows", reduce ? "play" : "pause");
  const resetBtn = button(bar, "Reset", "reset");
  resetBtn.hidden = true;
  const status = htmlEl("div", "sm-status", bar);
  status.setAttribute("aria-live", "polite");
  if (!data.cascade || !data.removed) {
    playBtn.disabled = true;
    playBtn.title = data.removed ? "This build of crm-trade-network does not report cascade rounds."
                                 : "The removed supplier has no position on the map.";
  }

  // ------------------------------------------------------------------ stage
  const W = data.map.width, H = data.map.height;
  const stage = htmlEl("div", "sm-stage", root);
  const svg = svgEl("svg", { viewBox: `0 0 ${W} ${H}`, role: "img", "aria-label": data.a11y }, stage);
  const gGrat = svgEl("g", { class: "grat" }, svg);
  const gLand = svgEl("g", { class: "land" }, svg);
  const gArcs = svgEl("g", { class: "arcs" }, svg);
  const gParticles = svgEl("g", { class: "particles" }, svg);
  const gBubbles = svgEl("g", { class: "bubbles" }, svg);
  const gFx = svgEl("g", { class: "fx" }, svg);
  const gLabels = svgEl("g", { class: "labels" }, svg);
  const tip = htmlEl("div", "sm-tip", stage);
  tip.setAttribute("role", "tooltip");

  data.map.graticule.forEach((d) => svgEl("path", { d }, gGrat));

  const landByCode = {};
  data.map.countries.forEach((c) => {
    const p = svgEl("path", { d: c.d }, gLand);
    c.codes.forEach((code) => { if (!landByCode[code]) landByCode[code] = p; });
    const code = c.codes.find((k) => data.info[k] || data.tint[k]);
    if (code) { p.dataset.a3 = code; p.classList.add("hot"); }
    const t = c.codes.map((k) => data.tint[k]).find((v) => v != null);
    if (t != null) p.style.fill = `rgba(${data.style.accentRgb},${t})`;
  });

  // ------------------------------------------------------------------ arcs + particles
  const arcs = data.flows.map((f, i) => {
    const el = svgEl("path", {
      d: `M${f.x1},${f.y1}Q${f.cx},${f.cy} ${f.x2},${f.y2}`, class: "arc", "stroke-width": f.w,
    }, gArcs);
    el.dataset.from = f.from;
    el.dataset.to = f.to;
    const len = el.getTotalLength();
    const g = svgEl("g", { class: "pgroup" }, gParticles);
    const n = Math.max(2, Math.min(7, Math.round(1.5 + f.w)));
    const size = 0.9 + f.w * 0.3;
    const dots = [];
    for (let j = 0; j < n; j++) {
      const halo = svgEl("circle", { r: size * 2.4, class: "p-halo" }, g);
      const t2 = svgEl("circle", { r: size * 0.55, class: "p-t2" }, g);
      const t1 = svgEl("circle", { r: size * 0.8, class: "p-t1" }, g);
      const head = svgEl("circle", { r: size, class: "p-head" }, g);
      dots.push({ phase: j / n, parts: [halo, t2, t1, head] });
    }
    // Speed follows arc length, so a long intercontinental flow takes longer to cross than a
    // short regional one; the bigger the flow, the more particles ride it.
    const period = 2.2 + len / 240;
    return { f, el, g, len, dots, period, i, cut: false };
  });

  // ------------------------------------------------------------------ circles + labels
  const bubbleEls = data.bubbles.map((b, i) => {
    const c = svgEl("circle", { cx: b.x, cy: b.y, r: Math.max(b.r, 2.2), class: "bubble", tabindex: "0" }, gBubbles);
    c.dataset.a3 = b.a3;
    c.setAttribute("aria-label", `${b.name}: ${b.value}${b.share ? ", " + b.share : ""}`);
    return { b, c, i };
  });
  // Labels for the largest producers. Each tries above its circle, then below, right and left,
  // and is left off rather than printed over another label; the tooltip still has the number.
  const placed = [];
  const clash = (a, b) => a.x < b.x + b.w && b.x < a.x + a.w && a.y < b.y + b.h && b.y < a.y + a.h;
  data.bubbles.slice(0, 6).forEach((b) => {
    const r = Math.max(b.r, 2.2);
    const t = svgEl("text", { class: "label" }, gLabels);
    t.textContent = b.name + " ";
    const v = svgEl("tspan", { class: "v" }, t);
    v.textContent = b.value;
    const spots = [[b.x, b.y - r - 5, "middle"], [b.x, b.y + r + 12, "middle"],
                   [b.x + r + 4, b.y + 4, "start"], [b.x - r - 4, b.y + 4, "end"]];
    for (const [x, y, anchor] of spots) {
      t.setAttribute("x", x);
      t.setAttribute("y", y);
      t.setAttribute("text-anchor", anchor);
      const bb = t.getBBox();
      const box = { x: bb.x - 2, y: bb.y - 1, w: bb.width + 4, h: bb.height + 2 };
      if (bb.x >= 0 && bb.x + bb.width <= W && bb.y >= 0 && !placed.some((p) => clash(p, box))) {
        placed.push(box);
        return;
      }
    }
    t.remove();
  });

  // ------------------------------------------------------------------ legend
  const legend = htmlEl("div", "sm-legend", root);
  const item = (svgMarkup, text) => {
    const it = htmlEl("div", "sm-li", legend);
    const s = svgEl("svg", { viewBox: "0 0 44 20", width: "44", height: "20", "aria-hidden": "true" }, it);
    s.innerHTML = svgMarkup;
    htmlEl("span", null, it, text);
  };
  item(`<circle cx="10" cy="10" r="4" class="bubble"/><circle cx="28" cy="10" r="8" class="bubble"/>`,
       `${data.text.measure} (${data.text.unit}); countries shaded by the same measure`);
  item(`<path d="M2 14Q22 0 42 14" class="arc" stroke-width="3"/><circle cx="30" cy="8" r="2.4" class="p-head"/>`,
       `${data.text.flows}; thicker is larger, dots move exporter to importer`);
  if (data.removed) {
    item(`<circle cx="22" cy="10" r="6" class="crit-dot"/>`, `${data.text.removedName}: the supplier removed in the disruption`);
  }

  // ------------------------------------------------------------------ focus + tooltip
  const infoText = (a3) => {
    const k = data.info[a3];
    if (!k) return null;
    const lines = [];
    if (k.mined) lines.push([data.text.measure, k.mined]);
    if (k.exports) lines.push(["Exports, 2023", k.exports]);
    if (k.imports) lines.push(["Imports, 2023", k.imports]);
    return { name: k.name, lines, buyers: k.buyers || [], hub: k.hub };
  };

  function showTip(a3, clientX, clientY) {
    const k = infoText(a3);
    if (!k) { tip.classList.remove("on"); return; }
    tip.replaceChildren();
    htmlEl("div", "t-name", tip, k.name);
    k.lines.forEach(([label, value]) => {
      const row = htmlEl("div", "t-row", tip);
      htmlEl("span", "t-k", row, label);
      htmlEl("span", "t-v", row, value);
    });
    if (k.buyers.length) {
      htmlEl("div", "t-k", tip, "Largest buyers");
      k.buyers.forEach((b) => htmlEl("div", "t-buyer", tip, b));
    }
    if (k.hub) htmlEl("div", "t-note", tip, "Re-export hub: part of this trade is transit, not origin.");
    const box = stage.getBoundingClientRect();
    tip.classList.add("on");
    const tw = tip.offsetWidth, th = tip.offsetHeight;
    let x = clientX - box.left + 14, y = clientY - box.top + 14;
    if (x + tw > box.width - 6) x = clientX - box.left - tw - 14;
    if (y + th > box.height - 6) y = Math.max(6, clientY - box.top - th - 14);
    tip.style.left = `${Math.max(6, x)}px`;
    tip.style.top = `${y}px`;
  }

  function focus(a3) {
    root.classList.toggle("focused", !!a3);
    arcs.forEach((a) => {
      const on = !a3 || a.f.from === a3 || a.f.to === a3;
      a.el.classList.toggle("dim", !on);
      a.g.classList.toggle("dim", !on);
    });
    bubbleEls.forEach(({ c, b }) => c.classList.toggle("dim", !!a3 && b.a3 !== a3));
  }

  const pointTarget = (ev) => ev.target.closest && ev.target.closest("[data-a3]");
  svg.addEventListener("pointermove", (ev) => {
    const t = pointTarget(ev);
    if (t) { showTip(t.dataset.a3, ev.clientX, ev.clientY); if (!pinned) focus(t.dataset.a3); }
    else { tip.classList.remove("on"); if (!pinned) focus(null); }
  });
  svg.addEventListener("pointerleave", () => { tip.classList.remove("on"); if (!pinned) focus(null); });
  svg.addEventListener("click", (ev) => {
    const t = pointTarget(ev);
    pinned = t && pinned !== t.dataset.a3 ? t.dataset.a3 : null;
    focus(pinned);
    status.textContent = pinned ? `Showing flows for ${(data.info[pinned] || {}).name || pinned}. Click it again to clear.` : "";
  });
  bubbleEls.forEach(({ c, b }) => {
    c.addEventListener("focus", () => { const r = c.getBoundingClientRect(); focus(b.a3); showTip(b.a3, r.right, r.top); });
    c.addEventListener("blur", () => { if (!pinned) focus(null); tip.classList.remove("on"); });
    c.addEventListener("keydown", (ev) => {
      if (ev.key === "Enter" || ev.key === " ") { ev.preventDefault(); pinned = pinned === b.a3 ? null : b.a3; focus(pinned); }
      if (ev.key === "Escape") { pinned = null; focus(null); tip.classList.remove("on"); }
    });
  });

  // ------------------------------------------------------------------ particle loop
  function frame(now) {
    raf = 0;
    if (!running) return;
    last = now;
    const clock = now / 1000;
    for (const a of arcs) {
      if (a.cut) continue;
      for (const d of a.dots) {
        const t = (clock / a.period + d.phase) % 1;
        const pts = [t, Math.max(0, t - 0.05), Math.max(0, t - 0.025), t];
        const fade = t < 0.08 ? t / 0.08 : t > 0.92 ? (1 - t) / 0.08 : 1;
        d.parts.forEach((el, k) => {
          const [x, y] = bez(a.f, pts[k]);
          el.setAttribute("cx", x.toFixed(1));
          el.setAttribute("cy", y.toFixed(1));
          el.style.opacity = fade;
        });
      }
    }
    raf = requestAnimationFrame(frame);
  }

  function setRunning() {
    const want = flowsOn && visible && !document.hidden;
    gParticles.style.display = flowsOn ? "" : "none";
    if (want && !running) { running = true; last = 0; raf = requestAnimationFrame(frame); }
    if (!want && running) { running = false; if (raf) cancelAnimationFrame(raf); raf = 0; }
  }

  flowBtn.addEventListener("click", () => {
    flowsOn = !flowsOn;
    flowBtn.setIcon(flowsOn ? "pause" : "play");
    flowBtn.setLabel(flowsOn ? "Pause flows" : "Show moving flows");
    setRunning();
  });
  const io = new IntersectionObserver((entries) => { visible = entries[0].isIntersecting; setRunning(); });
  io.observe(stage);
  const onVis = () => setRunning();
  document.addEventListener("visibilitychange", onVis);

  // ------------------------------------------------------------------ entrance
  const ease = "cubic-bezier(.22,.8,.26,1)";
  if (!reduce) {
    gLand.animate([{ opacity: 0 }, { opacity: 1 }], { duration: 600, easing: "ease-out" });
    gGrat.animate([{ opacity: 0 }, { opacity: 1 }], { duration: 900, easing: "ease-out" });
    bubbleEls.forEach(({ c, i }) => c.animate(
      [{ transform: "scale(0)", opacity: 0 }, { transform: "scale(1)", opacity: 1 }],
      { duration: 750, delay: 250 + i * 70, easing: "cubic-bezier(.34,1.56,.64,1)", fill: "backwards" }));
    arcs.forEach((a) => {
      a.el.style.strokeDasharray = `${a.len}`;
      const anim = a.el.animate([{ strokeDashoffset: a.len }, { strokeDashoffset: 0 }],
        { duration: 1100, delay: 500 + a.i * 40, easing: ease, fill: "backwards" });
      anim.onfinish = () => { a.el.style.strokeDasharray = ""; };
    });
    gParticles.style.opacity = 0;
    later(() => { gParticles.animate([{ opacity: 0 }, { opacity: 1 }], { duration: 600, fill: "forwards" }); }, 900 + arcs.length * 40);
    later(() => { gParticles.style.opacity = 1; }, 1600 + arcs.length * 40);
  }
  setRunning();

  // ------------------------------------------------------------------ disruption replay
  function ripple(x, y, r0, delay) {
    if (reduce) return;
    const c = svgEl("circle", { cx: x, cy: y, r: r0, class: "ripple" }, gFx);
    const grow = (r0 * 3.2 + 14) / r0;
    const a = c.animate([{ transform: "scale(1)", opacity: 0.85 }, { transform: `scale(${grow})`, opacity: 0 }],
      { duration: 1300, delay: delay || 0, easing: "cubic-bezier(.2,.6,.3,1)", fill: "both" });
    a.onfinish = () => c.remove();
  }

  function cutArcsTouching(a3) {
    arcs.forEach((a) => {
      if (!a.cut && (a.f.from === a3 || a.f.to === a3)) {
        a.cut = true;
        a.el.classList.add("cut");
        a.g.classList.add("cut");
      }
    });
  }

  function play() {
    if (played || !data.cascade || !data.removed) return;
    played = true;
    pinned = null;
    focus(null);
    playBtn.disabled = true;
    resetBtn.hidden = false;
    const R = data.removed;
    const step = reduce ? 450 : 1150;
    let down = 0;
    status.textContent = `${R.name} stops exporting.`;
    const src = bubbleEls.find(({ b }) => b.a3 === R.a3);
    if (src) src.c.classList.add("down");
    if (landByCode[R.a3]) landByCode[R.a3].classList.add("down");
    for (let k = 0; k < 3; k++) ripple(R.x, R.y, 6, k * 260);
    arcs.forEach((a) => {
      if (a.f.from !== R.a3) return;
      a.el.classList.add("severed");
      a.g.classList.add("cut");
      a.cut = true;
      if (!reduce) {
        a.el.style.strokeDasharray = "5 7";
        a.el.animate([{ strokeDashoffset: 0, opacity: 0.95 }, { strokeDashoffset: -60, opacity: 0.22 }],
          { duration: 1400, easing: "ease-in", fill: "forwards" });
      }
    });
    data.cascade.rounds.forEach((round, r) => {
      later(() => {
        round.countries.forEach((c, j) => {
          if (landByCode[c.a3]) landByCode[c.a3].classList.add("down");
          const bub = bubbleEls.find(({ b }) => b.a3 === c.a3);
          if (bub) bub.c.classList.add("down");
          ripple(c.x, c.y, 3, Math.min(j * 12, 500));
          cutArcsTouching(c.a3);
        });
        down += round.count;
        status.textContent = `Round ${r + 1}: ${round.count} more ${round.count === 1 ? "country" : "countries"} lose over `
          + `20% of their trade and fail (${down} of ${data.cascade.universe - 1} so far).`;
      }, step * (r + 1));
    });
    later(() => {
      const off = data.cascade.offmap ? ` ${data.cascade.offmap} of them have no position on this map.` : "";
      status.textContent = `Cascade over: losing ${R.name} takes down ${down} of the ${data.cascade.universe - 1} `
        + `other countries in this trade network within ${data.cascade.rounds.length + 1} rounds.${off}`;
    }, step * (data.cascade.rounds.length + 1) + 200);
  }

  function reset() {
    cleanup();
    parentElement.__smCleanup = null;
    mount(component);
  }

  playBtn.addEventListener("click", play);
  resetBtn.addEventListener("click", reset);

  // Labels only where they can be read: below ~560 px wide they would be a few pixels tall.
  const ro = new ResizeObserver(() => root.classList.toggle("compact", stage.clientWidth < 560));
  ro.observe(stage);

  function cleanup() {
    running = false;
    if (raf) cancelAnimationFrame(raf);
    timers.forEach(clearTimeout);
    io.disconnect();
    ro.disconnect();
    document.removeEventListener("visibilitychange", onVis);
    root.remove();
  }
  parentElement.__smCleanup = cleanup;
  return cleanup;
}

File: ui/supply_map.py
"""The Supply map page's data: where a material is mined, where it is traded, and how a
disruption spreads, prepared in Python so the browser only has to animate it.

Three layers, each from data this project already verifies elsewhere:
  - mine production or reserves by country: data/mine_supply.csv (USGS Mineral Commodity
    Summaries 2025 via the Materials Data Series, see scripts/build_mine_supply.py);
  - bilateral trade flows: the same 2023 UN Comtrade edge list every risk page uses;
  - the cascade replay: crm-trade-network's simulate_cascade, round by round.

Geometry (data/world_map.json) is projected once by scripts/build_world_map.py. Countries are
matched by ISO3 code: trade partners through UN Comtrade's own partner table, producers through
the explicit table in build_mine_supply.py. Anything that cannot be placed is counted and
reported on the page, never silently dropped.
"""
from __future__ import annotations

import json
import math
from pathlib import Path

import pandas as pd
import streamlit as st

from ui.theme import CRITICAL, SLATE

ROOT = Path(__file__).resolve().parent.parent
WORLD_FILE = ROOT / "data" / "world_map.json"
SUPPLY_FILE = ROOT / "data" / "mine_supply.csv"

# UN Comtrade reports Taiwan's trade as "Other Asia, nes" (code 490); every other name maps
# through the ISO3 code in Comtrade's own partner table.
COMTRADE_OVERRIDES = {"Other Asia, nes": "TWN"}

MAX_BUBBLE = 26.0      # SVG units, on a map 1000 units wide
MIN_FLOW_W, MAX_FLOW_W = 0.7, 6.0


@st.cache_data(show_spinner=False)
def load_world() -> dict:
    return json.loads(WORLD_FILE.read_text(encoding="utf-8"))


@st.cache_data(show_spinner=False)
def load_supply() -> pd.DataFrame:
    return pd.read_csv(SUPPLY_FILE, keep_default_na=False)


@st.cache_data(show_spinner=False)
def comtrade_iso3(partners_json: str) -> dict[str, str]:
    rows = json.loads(Path(partners_json).read_text(encoding="utf-8"))["results"]
    out = {r["PartnerDesc"]: r.get("PartnerCodeIsoAlpha3") or "" for r in rows}
    out.update(COMTRADE_OVERRIDES)
    return out


def _fmt_t(t: float) -> str:
    if t >= 1e6:
        return f"{t / 1e6:,.2f} Mt"
    if t >= 1e3:
        return f"{t / 1e3:,.1f} kt"
    return f"{t:,.0f} t"


def _fmt_usd(v: float) -> str:
    if v >= 1e9:
        return f"${v / 1e9:,.2f} bn"
    if v >= 1e6:
        return f"${v / 1e6:,.1f} M"
    return f"${v:,.0f}"


def _arc(x1: float, y1: float, x2: float, y2: float, height: float) -> tuple[float, float]:
    """Control point of a quadratic arc that bows away from the equator side, so flows read
    as arcs over the map rather than straight lines through it, and two opposite flows
    between the same pair of countries do not sit on top of each other."""
    mx, my = (x1 + x2) / 2, (y1 + y2) / 2
    dx, dy = x2 - x1, y2 - y1
    dist = math.hypot(dx, dy) or 1.0
    nx, ny = dy / dist, -dx / dist             # left-hand normal of the travel direction
    bend = min(0.24 * dist, 120.0)
    cx, cy = mx + nx * bend, my + ny * bend
    return round(min(max(cx, 2.0), 998.0), 1), round(min(max(cy, 2.0), height - 2.0), 1)


def build_payload(*, family: str, measure: str, edges: pd.DataFrame, top_n: int,
                  removed_country: str, cascade: dict, accent: str, commodity: str,
                  hs_code: str, partners_json: str, hubs: set[str], universe: int) -> tuple[dict, dict]:
    """Everything the map component draws, plus notes for the page (what could not be placed)."""
    world = load_world()
    anchors = world["anchors"]
    iso = comtrade_iso3(partners_json)

    def place(name: str) -> tuple[str, list[float]] | None:
        code = iso.get(name, "")
        return (code, anchors[code]) if code in anchors else None

    # ---- production / reserves circles
    rows = load_supply()
    rows = rows[(rows.family == family) & (rows.measure == measure)]
    placed_rows = rows[rows.iso3 != ""]
    other_rows = rows[rows.iso3 == ""]
    vmax = float(placed_rows.tonnes.max()) if len(placed_rows) else 1.0
    bubbles, tint = [], {}
    for r in placed_rows.sort_values("tonnes", ascending=False).itertuples():
        if r.iso3 not in anchors:
            continue
        x, y = anchors[r.iso3]
        share = float(r.share_pct) if str(r.share_pct).strip() else None
        bubbles.append({
            "a3": r.iso3, "name": r.country, "x": x, "y": y,
            "r": round(MAX_BUBBLE * math.sqrt(r.tonnes / vmax), 2),
            "value": _fmt_t(r.tonnes), "share": f"{share:.0f}% of world" if share is not None else "",
        })
        tint[r.iso3] = round(0.10 + 0.32 * math.sqrt(r.tonnes / vmax), 3)
    unit = rows.unit.iloc[0] if len(rows) else ""
    year = int(rows.year.iloc[0]) if len(rows) else None

    # ---- trade flows: the largest links, plus the removed supplier's own largest ones so the
    # disruption replay always has its arcs on the map
    e = edges[edges.exporter != edges.importer].sort_values("value_usd", ascending=False)
    chosen = e.head(top_n)
    own = e[e.exporter == removed_country].head(8)
    chosen = pd.concat([chosen, own]).drop_duplicates(subset=["exporter", "importer"])
    vmax_f = float(chosen.value_usd.max()) if len(chosen) else 1.0
    flows, unplaced = [], set()
    for r in chosen.itertuples():
        a, b = place(r.exporter), place(r.importer)
        if not a or not b:
            unplaced.update(n for n, p in ((r.exporter, a), (r.importer, b)) if not p)
            continue
        (fa3, (x1, y1)), (ta3, (x2, y2)) = a, b
        cx, cy = _arc(x1, y1, x2, y2, world["height"])
        flows.append({
            "from": fa3, "to": ta3, "fromName": r.exporter, "toName": r.importer,
            "x1": x1, "y1": y1, "x2": x2, "y2": y2, "cx": cx, "cy": cy,
            "w": round(MIN_FLOW_W + (MAX_FLOW_W - MIN_FLOW_W) * math.sqrt(r.value_usd / vmax_f), 2),
            "usd": _fmt_usd(r.value_usd), "tonnes": _fmt_t(r.net_kg / 1000) if r.net_kg > 0 else "n/a",
            "removed": r.exporter == removed_country,
        })

    # ---- per-country tooltip facts, from the full edge list (not only the arcs drawn)
    info = {}
    by_exp = edges.groupby("exporter").value_usd.sum()
    by_imp = edges.groupby("importer").value_usd.sum()
    total_trade = float(edges.value_usd.sum()) or 1.0
    names_on_map = {f["fromName"] for f in flows} | {f["toName"] for f in flows} | {removed_country}
    for name in names_on_map:
        p = place(name)
        if not p:
            continue
        partners = (edges[edges.exporter == name].groupby("importer").value_usd.sum()
                    .sort_values(ascending=False).head(3))
        info[p[0]] = {
            "name": "Taiwan (reported as \"Other Asia, nes\")" if name == "Other Asia, nes" else name,
            "exports": f"{_fmt_usd(float(by_exp.get(name, 0.0)))} · {by_exp.get(name, 0.0) / total_trade:.1%} of world trade",
            "imports": _fmt_usd(float(by_imp.get(name, 0.0))),
            "buyers": [f"{k} {_fmt_usd(float(v))}" for k, v in partners.items()],
            "hub": name in hubs,
        }
    for b in bubbles:
        info.setdefault(b["a3"], {"name": b["name"]})
        info[b["a3"]]["mined"] = f"{b['value']}" + (f" · {b['share']}" if b["share"] else "")

    # ---- cascade replay, round by round
    history = cascade.get("history")
    rounds, offmap = [], 0
    if history is not None:
        for wave in history:
            placed_wave = []
            for name in wave:
                p = place(name)
                if p:
                    placed_wave.append({"a3": p[0], "name": name, "x": p[1][0], "y": p[1][1]})
                else:
                    offmap += 1
            rounds.append({"countries": placed_wave, "count": len(wave)})
    removed = place(removed_country)

    h = accent.lstrip("#")
    payload = {
        "map": {"width": world["width"], "height": world["height"], "graticule": world["graticule"],
                "countries": [{"codes": c["codes"], "d": c["d"]} for c in world["countries"]]},
        "style": {"accent": accent, "accentRgb": ",".join(str(int(h[i:i + 2], 16)) for i in (0, 2, 4)),
                  "critical": CRITICAL, "slate": SLATE},
        "tint": tint, "bubbles": bubbles, "flows": flows, "info": info,
        "removed": ({"a3": removed[0], "name": removed_country, "x": removed[1][0], "y": removed[1][1]}
                    if removed else None),
        "cascade": ({"rounds": rounds, "universe": universe, "offmap": offmap}
                    if history is not None else None),
        "text": {
            "measure": f"{measure}{f', {year}' if year else ''}", "unit": unit,
            "flows": f"Trade in {commodity} (HS {hs_code}), 2023, by value",
            "removedName": removed_country,
        },
        "a11y": (f"World map of {family.lower()}: circles sized by {measure.lower()}, arcs showing the "
                 f"{len(flows)} largest trade flows of {commodity} in 2023."),
    }
    notes = {
        "unplaced_trade": sorted(unplaced),
        "other_supply": [(r.country, _fmt_t(r.tonnes)) for r in other_rows.itertuples()],
        "flows_drawn": len(flows), "bubbles": len(bubbles), "unit": unit, "year": year,
        "cascade_offmap": offmap, "has_history": history is not None,
    }
    return payload, notes


@st.cache_resource(show_spinner=False)
def _component_source() -> tuple[str, str]:
    here = Path(__file__).resolve().parent
    return ((here / "supply_map.css").read_text(encoding="utf-8"),
            (here / "supply_map.js").read_text(encoding="utf-8"))


def supply_map(payload: dict, key: str):
    """Mount the animated map. Registered on every run, the pattern Streamlit's own docs use:
    the component registry is rebuilt per script run, so a definition cached across runs goes
    stale ("Component is not registered") from the second run on."""
    css, js = _component_source()
    component = st.components.v2.component("mfa_supply_map", html='<div class="sm-mount"></div>', css=css, js=js)
    return component(data=payload, key=key, width="stretch", height="content")
